keps/sig-storage/3314-csi-volume-snapshot-delta/README.md

```diff
@@ -0,0 +1,1192 @@
+architecture.
+* Support provider-specific user input parameters such as block size,
+fixed-sized vs. variable-sized blocks etc., without leaking provider-specific
+implementation into CSI.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+* Retrieval of the actual data blocks is outside the scope of this KEP. The
+proposed API returns only the metadata of the changed blocks.
+* The ability to identify changes on file snapshots are slated for future KEP.
+This includes support for directories, subdirectories and files.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+This KEP introduces a new CRD called `VolumeSnapshotDelta` to the CSI
+architecture.
+
+The CRD abstracts away the details around interacting with the storage
+providers' CBT endpoints. Essentially, the CRD allows a Kubernetes user to say,
+
+> Help me find all the data blocks that have changed between these two
+> snapshots.
+
+The new `VolumeSnapshotDelta` component must be able to handle large amount of
+data returned by the storage providers, without negatively impacting the rest of
+the cluster through resource contention and starvation. Specifically, this KEP
+emphasizes on not putting Kubernetes' etcd in the CBT datapath, to avoid bogging
+it down with heavy IOPS operations.
+
+The `VolumeSnapshotDelta` component will be implemented as a [CSI sidecar]
+container, following the existing CSI driver deployment model. The new component
+will have:
+
+* a controller that watches for new `VolumeSnapshotDelta` resources.
+* a HTTP listener the provides direct datapaths to handle CBT requests and
+responses.
+
+[CSI sidecar]: https://kubernetes-csi.github.io/docs/sidecar-containers.html
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Segmenting Critical Datapath Resource Requirements
+
+It's not uncommon that the response payload from the storage providers can be
+quite large, in the magnitude of gigabytes of changed block metadata, per volume
+snapshot pair. The usual declarative approach of updating the custom
+resource's `status` subresource with the payload may put etcd at risk of
+storage capacity exhaustion as well as regular IOPS spikes.
+
+To mitigate this issue, this KEP proposes a 2-hop request mechanism where
```

Comment:
Yeah, we did do an alternate prototype with aggregated api server. The main issue there it's that it will put K8s api server in the request/response path. We thought for alpha, perhaps we can postpone an aggregated api server, until it's truly necessary. Getting inputs from the api machinery group is a good idea. Do you see this as a blocker for this KEP?

Reply:
Not from me, that's up to SIG storage :)

Reply:
I don't know much about aggregate API server, so I might be completed y But each CSI driver runs its own sidecars and there may be multiple CSI drivers running in the same cluster.

With this KEP, each driver can run its own VolumeSnapshotDelta sidecar, with its own HTTP(S) endpoint. Each sidecar watches all VolumeSnapshotDelta CRs, filters them by VolumeSnapshot driver name, and reacts only to CRs that are for the driver they run with.

If each CSI driver runs its own aggregated API server, how can all of the aggregated APi servers provide **the same** VolumeSnapshotDelta objects, filtering by CSI driver name?

Reply:
If run as a sidecar, the aggregated apiserver should still be able to distinguish the CRs using the same driver-name approach, as the current proposed approach.

Alternately, I am thinking if we can run the aggregated apiserver as its own separate service. It discovers the CSI drivers endpoints via a driver-to-cbt-endpoint mapping, stored either in a configmap or the `CSIDriver` resources. Is this a viable approach?

Reply:
So, there would be a common snapshot-apiserver, just like snapshot-controller, that would provide Kubernetes API to all CSI drivers:

```
       backup app 1      backup app 2
                │            │
                │            │
                └─────┬──────┘
                      │
                      ▼
              snapshot-apiserver
                      │
                      │
                ┌─────┴───────┐
                │             │
                ▼             ▼
     CSI driver 1             CSI driver 2
external-delta-sidecar   external-delta-sidecar
```

And the actual CBD metadata would be available as a subresource, like say `VolumeSnapshotDelta/changedblocks`. That is possible. Question is, if it's worth the effort. I can see a benefit in slightly nicer RBAC - you can set RBAC for a subresource, similarly as we have for `/status` of all objects. But the complexity and mount of code to maintain would grow significantly.